source/grammar/qasm3.g4

```diff
@@ -1,98 +1,461 @@
     ;
 
-Gate
-    : 'gate'
+call
+    : Identifier
+    | builtInMath
+    | castOperator
     ;
 
-Assign : '=';
+builtInMath
+    : 'sin' | 'cos' | 'tan' | 'exp' | 'ln' | 'sqrt' | 'popcount' | 'lengthof'
+    ;
 
-StringLiteral
-    : '"' Identifier '"'
+castOperator
+    : classicalType
     ;
 
-OpenBracket : '[';
+incrementor
+    : '++'
+    | '--'
+    ;
 
-CloseBracket : ']';
+assignmentExpression
+    : assignmentOperator expression
+    ;
 
-Colon: ':';
+assignmentOperator
+    : ASSIGN
+    | ARROW
+    | '+=' | '-=' | '*=' | '/=' | '&=' | '|=' | '~=' | '^=' | '<<=' | '>>='
+    ;
 
-SemiColon : ';';
+membershipTest
+    : Identifier 'in' setDeclaration
+    ;
 
-Dot : '.';
+setDeclaration
+    : LBRACE expressionList RBRACE
+    | range
+    ;
 
-Comma : ',';
+loopBranchBlock
+    : statement
+    | programBlock
+    ;
+
+branchingStatement
+    : 'if' LPAREN expression RPAREN loopBranchBlock ( 'else' loopBranchBlock )?
+    ;
+
+loopStatement: ( 'for' membershipTest | 'while' LPAREN expression RPAREN ) loopBranchBlock
+    ;
+
+controlDirectiveStatement
+    : controlDirective SEMICOLON
+    ;
+
+controlDirective
+    : 'break'
+    | 'continue'
+    | 'end'
+    ;
+
+kernelDeclaration
+    : 'kernel' Identifier ( LPAREN classicalTypeList? RPAREN )? returnSignature?
+    classicalType? SEMICOLON
+    ;
+
+/* Subroutines */
+
+subroutineDefinition
+    : 'def' Identifier ( LPAREN subroutineArgumentList? RPAREN )? returnSignature? programBlock
+    ;
+
+subroutineArgumentList
+    : classicalArgumentList | quantumArgumentList
+    ;
+
+/* Directives */
+
+pragma
+    : '#pragma' LPAREN AnyString RPAREN
+    ;
+
+/* Circuit Timing */
+
+timingBox
+    : 'boxas' Identifier quantumBlock
+    | 'boxto' timeUnit quantumBlock
+    ;
+
+timeTerminator
+    : timeIdentifier | 'stretchinf'
+    ;
+
+timeIdentifier
+    : Identifier timeUnit?
+    | 'lengthof' LPAREN Identifier RPAREN
+    ;
+
+timeUnit
+    : 'dt' | 'ns' | 'us' | 'ms' | 's' ;
+
+/* Pulse Level Descriptions of Gates and Measurement */
+
+physicalQubitIdentifier
+    : '%' 'q'? Integer
+    ;
+
+physicalQubitList
+    : ( physicalQubitIdentifier COMMA )* physicalQubitIdentifier
+    ;
+
+calibration
+    : calibrationGrammarDeclaration
+    | calibrationDefinition
+    ;
+
+calibrationGrammarDeclaration
+    : 'defcalgrammar' calibrationGrammar SEMICOLON
+    ;
+
+calibrationDefinition
+    : 'defcal' calibrationGrammar? Identifier
+    ( LPAREN calibrationArgumentList? RPAREN )? physicalQubitList
+    returnSignature calibrationBody
+    ;
+
+calibrationGrammar
+    : ( 'openpulse' | Identifier )*
+    ;
+
+calibrationArgumentList
+    : classicalArgumentList | expressionList
+    ;
+
+calibrationBody
+    : LBRACE AnyString RBRACE
+    ;
+
+/** Lexer grammar **/
+LBRACKET : '[' ;
+RBRACKET : ']' ;
+
+LBRACE : '{' ;
+RBRACE : '}' ;
+
+LPAREN : '(' ;
+RPAREN : ')' ;
+
+COLON: ':' ;
+SEMICOLON : ';' ;
+
+DOT : '.' ;
+COMMA : ',' ;
+
+AnyString : .*? ;
```

Comment:
warning: non-fragment lexer rule AnyString can match the empty string

Reply:
Addressed in https://github.com/Qiskit/openqasm/pull/89/commits/c3b3b28409aec991623c590a5ecfe28c7527d14c.